### PR TITLE
boot.loader.grub: add theme option

### PR DESCRIPTION
**Pull request by mkg20001**
###### Motivation for this change

Added theme option for grub

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mkg20001** [comment]
@yutyo this is the thing you need

File: nixos/modules/system/boot/loader/grub/install-grub.pl
Comment (mkg20001) [inline nixos/modules/system/boot/loader/grub/install-grub.pl:378]: I want to get this hack removed before merge, anyone with good perl knowledge has any idea?

Comment (mkg20001) [reply nixos/modules/system/boot/loader/grub/install-grub.pl:389]: Breaks if the .pf2 file is in a subfolder of the theme, since we're referencing it by it's filename, not sub-path

File: nixos/modules/system/boot/loader/grub/install-grub.pl
Comment (samueldr) [inline nixos/modules/system/boot/loader/grub/install-grub.pl:384]: Is it an issue with grub, though ugly, if the path starts with two slashes?

If it isn't we should drop the unnecessary ternary, so the code is easier to understand.

Alternatively, there may be something perl-y that handles paths concatenation for us, rather than doing the hard task ourselves?

File: nixos/modules/system/boot/loader/grub/install-grub.pl
Comment (samueldr) [reply nixos/modules/system/boot/loader/grub/install-grub.pl:389]: > The wanted function takes no arguments but rather does its work through a collection of variables.
> 
>     `$File::Find::dir` is the current directory name,
>     `$_` is the current filename within that directory
>     `$File::Find::name` is the complete pathname to the file.
> 

https://perldoc.perl.org/File/Find.html#The-wanted-function

I believe you should be able to replace `$_` with `$File::Find::name` to fix the issue with files in subfolders.

Comment (samueldr) [reply nixos/modules/system/boot/loader/grub/install-grub.pl:389]: Alternatively, could we instead leave it to the user to set `font` separately from `theme`?

It's not like it's a self-contained theme system anyway, which is why we have this hack.

Having distinct options allows more customization in the end. The user would use `"${pkgs.nixos-grub-theme}/the/font.pf2"` as a value in that instance.

Or am I missing something?

**Review by samueldr — COMMENTED** [review]
Looks okay, though not big on perl, I don't see problems.

Though obviously, approval only once the discussion about the "hacky" function resolved.

File: nixos/modules/system/boot/loader/grub/install-grub.pl
Comment (mkg20001) [reply nixos/modules/system/boot/loader/grub/install-grub.pl:389]: The theme specifies the font, we're just loading it

See ${nixos-grub2-theme}/theme.txt

File: nixos/modules/system/boot/loader/grub/install-grub.pl
Comment (mkg20001) [reply nixos/modules/system/boot/loader/grub/install-grub.pl:384]: I copied this from the other code, just following the pattern here... not my department :P

https://github.com/NixOS/nixpkgs/pull/83678/files#diff-3f0e9813c2baa6a4e7f8dbe5fd38ea56L364

File: nixos/modules/system/boot/loader/grub/install-grub.pl
Comment (samueldr) [reply nixos/modules/system/boot/loader/grub/install-grub.pl:389]: Right, [I see what I did there](https://github.com/NixOS/nixpkgs/blob/82a441d3cfcccce66500e40358a241ed9ccb7c47/nixos/modules/installer/cd-dvd/iso-image.nix#L204-L205).

It's been such a long time I don't remember all that :|

I figure then that this `find` is appropriate, given that AFAICT from superficial searches there is no recursive glob. That's the perl way to do it.

File: nixos/modules/system/boot/loader/grub/grub.nix
Comment (edolstra) [inline nixos/modules/system/boot/loader/grub/grub.nix:435]:
```suggestion
          Grub theme to be used.
```

File: nixos/modules/system/boot/loader/grub/grub.nix
Comment (edolstra) [inline nixos/modules/system/boot/loader/grub/grub.nix:432]:
```suggestion
        example = literalExample "[ pkgs.nixos-grub2-theme ]";
```

**Review by edolstra — APPROVED** [review]

File: nixos/modules/system/boot/loader/grub/install-grub.pl
Comment (samueldr) [reply nixos/modules/system/boot/loader/grub/install-grub.pl:384]: Ah, then don't worry too much, even though the question still stands in a more global manner.

File: nixos/modules/system/boot/loader/grub/install-grub.pl
Comment (samueldr) [inline nixos/modules/system/boot/loader/grub/install-grub.pl:395]: I'm not sure if you're currently working on it or not, but for the previously stated issue about non-top-level font files:


> > The wanted function takes no arguments but rather does its work through a collection of variables.
> > 
> >     `$File::Find::dir` is the current directory name,
> >     `$_` is the current filename within that directory
> >     `$File::Find::name` is the complete pathname to the file.
> > 
> 
> https://perldoc.perl.org/File/Find.html#The-wanted-function
> 
> I believe you should be able to replace `$_` with `$File::Find::name` to fix the issue with files in subfolders.

**Comment by mkg20001** [comment]
merge?

File: nixos/modules/system/boot/loader/grub/grub.nix
Comment (samueldr) [inline nixos/modules/system/boot/loader/grub/grub.nix:432]: I was *testing* the PR before clicking the merge button, rather than only reading.

It's not accepting a list of paths, but only one path.

```suggestion
        example = literalExample "pkgs.nixos-grub2-theme";
```

**Review by samueldr — CHANGES_REQUESTED** [review]
I have tested the PR and, other than this *documentation bug*, it seems to work just fine.

So once this minor bug squashed, it's ready for a merge.

image

**Comment by samueldr** [comment]
Thank you very much!

**Pull request merged**
